<a href="https://colab.research.google.com/github/LegendSeyi/Thai_Road_Accident_Data_Analysis/blob/main/Thai_Road_Accident_data_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **THAI ROAD ACCIDENT ANALYSIS**
#### **_By Adeyemi Oluwaseyi Emmanuel_**

[Click here to check Linkedin profile](https://www.linkedin.com/in/oluwaseyi-adeyemi-33b1ab197/)

[Twitter link](https://twitter.com/AmLegendseyi)

## **INTRODUCTION**

This dataset provides comprehensive statistics on recorded road accidents in Thailand, spanning from approximately 2019 to 2022. The data was sourced from raw information provided by the Office of the Permanent Secretary, Ministry of Transport, which is also utilized in this public dashboard for easier access and visualization. The dataset encompasses various aspects of road accidents and aims to shed light on the trends and patterns within this critical area of concern, analysis of this data could be crucial in guiding road safety policies and measures👍.

We import our required libraries

In [44]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [45]:
## creating a dictionary in python and converting it to a dataframe
tb = {'Column':['acc_code',
    'incident_datetime',
    'report_datetime',
    'province_th',
    'province_en',
    'agency',
    'route',
    'vehicle_type',
    'presumed_cause',
    'accident_type',
    'number_of_vehicles_involved',
    'number_of_fatalities',
    'number_of_injuries',
    'weather_condition',
    'latitude',
    'longitude',
    'road_description',
    'slope_description'],
    'Description':['The accident code or identifier',
    'The date and time of the accident occurrence',
    'The date and time when the accident was reported',
    'The name of the province in Thailand, written in Thai',
    'The name of the province in Thailand, written in English',
    'The government agency responsible for the road and traffic management',
    'The route or road segment where the accident occurred',
    'The type of vehicle involved in the accident',
    'The presumed cause or reason for the accident',
    'The type or nature of the accident',
    'The number of vehicles involved in the accident',
    'The number of fatalities resulting from the accident',
    'The number of injuries resulting from the accident',
    'The weather condition at the time of the accident',
    'The latitude coordinate of the accident location',
    'The longitude coordinate of the accident location',
    'The description of the road type or configuration where the accident occurred',
    'The description of the slope condition at the accident location']
     }

In [46]:
## Converting the dictionary to a data frame
dico = pd.DataFrame(tb)

In [47]:
dico

Column  \
0                      acc_code   
1             incident_datetime   
2               report_datetime   
3                   province_th   
4                   province_en   
5                        agency   
6                         route   
7                  vehicle_type   
8                presumed_cause   
9                 accident_type   
10  number_of_vehicles_involved   
11         number_of_fatalities   
12           number_of_injuries   
13            weather_condition   
14                     latitude   
15                    longitude   
16             road_description   
17            slope_description   

                                          Description  
0                     The accident code or identifier  
1        The date and time of the accident occurrence  
2    The date and time when the accident was reported  
3   The name of the province in Thailand, written ...  
4   The name of the province in Thailand, written ...  
5   The government agency responsible for the road...  
6   The route or road segment where the accident o...  
7        The type of vehicle involved in the accident  
8       The presumed cause or reason for the accident  
9                  The type or nature of the accident  
10    The number of vehicles involved in the accident  
11  The number of fatalities resulting from the ac...  
12  The number of injuries resulting from the acci...  
13  The weather condition at the time of the accident  
14   The latitude coordinate of the accident location  
15  The longitude coordinate of the accident location  
16  The description of the road type or configurat...  
17  The description of the slope condition at the ...

The above table is the description of each columns in the dataset....

With this information, we can now import our daraset for the analysis

## **UNDERSTANDING THE DATASET**

In [48]:
# We will use the pandas library to import our dataset

In [ ]:
# We can import our dataset from github


url = 'https://raw.githubusercontent.com/LegendSeyi/Dataset/main/thai_road_accident_2019_2022.csv'

pd.set_option('display.max_columns', None)
data = pd.read_csv(url, encoding= 'utf-8')

In [50]:
data.head()

acc_code    incident_datetime      report_datetime      province_th  \
0    571905  2019-01-01 00:00:00  2019-01-02 06:11:00           ลพบุรี   
1   3790870  2019-01-01 00:03:00  2020-02-20 13:48:00      อุบลราชธานี   
2    599075  2019-01-01 00:05:00  2019-01-01 10:35:00  ประจวบคีรีขันธ์   
3    571924  2019-01-01 00:20:00  2019-01-02 05:12:00        เชียงใหม่   
4    599523  2019-01-01 00:25:00  2019-01-04 09:42:00        นครสวรรค์   

           province_en                     agency  \
0               Loburi  department of rural roads   
1     Ubon Ratchathani     department of highways   
2  Prachuap Khiri Khan     department of highways   
3           Chiang Mai  department of rural roads   
4         Nakhon Sawan     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเลข 21 (กม.ที่ 31+000) - บ้านวัง...             motorcycle   
1                              เดชอุดม - อุบลราชธานี  private/passenger car   
2                              ปราณบุรี - ปากน้ำปราณ             motorcycle   
3  เชื่อมทางหลวงหมายเลข 1013 (กม.ที่ 8+200) - บ้า...             motorcycle   
4                                   เกยไชย - ศรีมงคล  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

We have imported our data, so we can check the size of the dataset

In [51]:
data.shape

(81735, 18)

We have 81735 row and 18 columns in the dataset

In [52]:
data.columns

Index(['acc_code', 'incident_datetime', 'report_datetime', 'province_th',
       'province_en', 'agency', 'route', 'vehicle_type', 'presumed_cause',
       'accident_type', 'number_of_vehicles_involved', 'number_of_fatalities',
       'number_of_injuries', 'weather_condition', 'latitude', 'longitude',
       'road_description', 'slope_description'],
      dtype='object')

we generated the list of columns available, but out first column is not properly named..

So we have successfully renamed the column

Generate the information of the dataset

In [53]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 81735 entries, 0 to 81734
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   acc_code                     81735 non-null  int64  
 1   incident_datetime            81735 non-null  object 
 2   report_datetime              81735 non-null  object 
 3   province_th                  81735 non-null  object 
 4   province_en                  81735 non-null  object 
 5   agency                       81735 non-null  object 
 6   route                        81735 non-null  object 
 7   vehicle_type                 81735 non-null  object 
 8   presumed_cause               81735 non-null  object 
 9   accident_type                81735 non-null  object 
 10  number_of_vehicles_involved  81735 non-null  int64  
 11  number_of_fatalities         81735 non-null  int64  
 12  number_of_injuries           81735 non-null  int64  
 13  weather_conditio

We can observe that we have different data types such ad float, int, object

In [54]:
# code to generate the number of null values present in the columns
data.isnull().sum()

acc_code                         0
incident_datetime                0
report_datetime                  0
province_th                      0
province_en                      0
agency                           0
route                            0
vehicle_type                     0
presumed_cause                   0
accident_type                    0
number_of_vehicles_involved      0
number_of_fatalities             0
number_of_injuries               0
weather_condition                0
latitude                       359
longitude                      359
road_description                 0
slope_description                0
dtype: int64

we can observe that column latitude and longitude contain the same number of missing values

## **DATA CLEANING**

### **Incident_datetime column**

In [55]:
# We start by analysing the 'incident datetime' column , we convert the type of column into a datetime
#data['incident_datetime'] = pd.to_datetime(data['incident_datetime'])

In [56]:
data['incident_datetime'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 81735 entries, 0 to 81734
Series name: incident_datetime
Non-Null Count  Dtype 
--------------  ----- 
81735 non-null  object
dtypes: object(1)
memory usage: 638.7+ KB


The column has been converted to a datetime column. so lets check if there are no null values in the dataset

In [57]:
data['incident_datetime'].isnull().sum()

0

We can confirm that there are no null values.

We can sort the data set with the incident datetime column

In [58]:
data.sort_values(by='incident_datetime', inplace=True)

Now we can be sure that the dataset has been sorted

**Now we want to create two different column from the datetime column, one of the date (Year-Month-Day) and another for the time zone (Hour-minute-Seconds)**

In [59]:
# Creating a new column of just the date
# Creating a new column of just the time
data['incident_date'] = data['incident_datetime'].str[:10]
data['incident_time'] = data['incident_datetime'].str[10:]


Now we can drop/del the incident_datetime column and also rearrange the data bring the newly created columns to the front

In [60]:

# we use the pop function
first_col = data.pop('incident_date')
second_col = data.pop('incident_time')

# then we insert it the column number position, the name of the columns, and the coloumn we popped
data.insert(1, 'Incident_date', first_col)
data.insert(2, 'Incident_time', second_col)


### **Report_datetime column**

Now we want to create two different column from the datetime column, one of the date (Year-Month-Day) and another for the time zone (Hour-minute-Seconds)

In [61]:
# Creating a new column of just the date
# Creating a new column of just the time
data['report_date'] = data['report_datetime'].str[:10]
data['report_time'] = data['report_datetime'].str[10:]

Now we can drop/del the incident_datetime column and also rearrange the data bring the newly created columns to the front

i use Ctrl + shift + L to multiple select a repeated string after highlighting of of it

In [62]:
# we use the pop function
first_col = data.pop('report_date')
second_col = data.pop('report_time')

# then we insert it the column number position, the name of the columns, and the coloumn we popped
data.insert(3, 'report_date', first_col)
data.insert(4, 'report_time', second_col)



In [63]:
# we can use the drop function to delete the report datetime column
data.drop(['report_datetime','incident_datetime'], axis=1, inplace=True)
data.head()


acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   
3    571924    2019-01-01      00:20:00  2019-01-02    05:12:00   
4    599523    2019-01-01      00:25:00  2019-01-04    09:42:00   

       province_th          province_en                     agency  \
0           ลพบุรี               Loburi  department of rural roads   
1      อุบลราชธานี     Ubon Ratchathani     department of highways   
2  ประจวบคีรีขันธ์  Prachuap Khiri Khan     department of highways   
3        เชียงใหม่           Chiang Mai  department of rural roads   
4        นครสวรรค์         Nakhon Sawan     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเลข 21 (กม.ที่ 31+000) - บ้านวัง...             motorcycle   
1                              เดชอุดม - อุบลราชธานี  private/passenger car   
2                              ปราณบุรี - ปากน้ำปราณ             motorcycle   
3  เชื่อมทางหลวงหมายเลข 1013 (กม.ที่ 8+200) - บ้า...             motorcycle   
4                                   เกยไชย - ศรีมงคล  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

**From our observation of the datetime columns, we noticed that in some rows there are difference in the reported  date/time and when the inccident occurred.... we can get some insight by analysing this as to why the delay occurred**

### **Province_th Column**

In [64]:
# preview the column
data['province_th']

0                 ลพบุรี
1            อุบลราชธานี
2        ประจวบคีรีขันธ์
3              เชียงใหม่
4              นครสวรรค์
              ...       
81730            นนทบุรี
81731               ตราด
81732          มหาสารคาม
81733          กำแพงเพชร
81734           เชียงราย
Name: province_th, Length: 81735, dtype: object

From the description, this column is the names of the provinces the incidents occurred..

I decided to drop to column, the texts were not well represented..

In [65]:
data.drop('province_th', axis=1, inplace=True)

### **Province_en Column**

In [66]:
#preview the column
data['province_en']

0                     Loburi
1           Ubon Ratchathani
2        Prachuap Khiri Khan
3                 Chiang Mai
4               Nakhon Sawan
                ...         
81730             Nonthaburi
81731                   Trat
81732          Maha Sarakham
81733         Kamphaeng Phet
81734             Chiang Rai
Name: province_en, Length: 81735, dtype: object

Lets check the number of province involved

In [67]:
# LIst of the unique province in the dataset
data['province_en'].unique()

array(['Loburi', 'Ubon Ratchathani', 'Prachuap Khiri Khan', 'Chiang Mai',
       'Nakhon Sawan', 'Mae Hong Son', 'Chumphon', 'Sing Buri',
       'Songkhla', 'Lamphun', 'Trat', 'Phuket', 'Saraburi', 'Ratchaburi',
       'Phra Nakhon Si Ayutthaya', 'Nakhon Ratchasima',
       'Nakhon Si Thammarat', 'Chaiyaphum', 'Kalasin', 'Suphan Buri',
       'Phetchaburi', 'Phrae', 'Chai Nat', 'Prachin Buri',
       'Nakhon Pathom', 'Kanchanaburi', 'Phetchabun', 'Ang Thong',
       'Nonthaburi', 'Samut Prakan', 'Bangkok', 'Phayao', 'Phatthalung',
       'Yala', 'Maha Sarakham', 'Surat Thani', 'Amnat Charoen',
       'Nong Khai', 'Nan', 'Phangnga', 'Narathiwat', 'Samut Sakhon',
       'Chanthaburi', 'Samut Songkhram', 'Phitsanulok', 'Pathum Thani',
       'Tak', 'Loei', 'Chiang Rai', 'Chachoengsao', 'Buri Ram',
       'Uthai Thani', 'Krabi', 'Surin', 'Udon Thani', 'Si Sa Ket',
       'Uttaradit', 'Khon Kaen', 'Kamphaeng Phet', 'Yasothon', 'Satun',
       'Nakhon Nayok', 'Chon Buri', 'Rayong', 'buogkan'

Let convert this column to a categorical column

In [68]:
data['province_en'] = data['province_en'].astype('category')

In [69]:
data['province_en'].info()

<class 'pandas.core.series.Series'>
Int64Index: 81735 entries, 0 to 81734
Series name: province_en
Non-Null Count  Dtype   
--------------  -----   
81735 non-null  category
dtypes: category(1)
memory usage: 721.0 KB


Renaming the province_en to province since we dropped the thai-meanning

In [70]:
data.columns = [col.replace("_en", "") for col in data.columns]

In [71]:
data.head(3)

acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   

              province                     agency  \
0               Loburi  department of rural roads   
1     Ubon Ratchathani     department of highways   
2  Prachuap Khiri Khan     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเลข 21 (กม.ที่ 31+000) - บ้านวัง...             motorcycle   
1                              เดชอุดม - อุบลราชธานี  private/passenger car   
2                              ปราณบุรี - ปากน้ำปราณ             motorcycle   

                           presumed_cause                       accident_type  \
0  driving under the influence of alcohol                               other   
1                                speeding    rollover/fallen on straight road   
2                                speeding  head-on collision (not overtaking)   

   number_of_vehicles_involved  number_of_fatalities  number_of_injuries  \
0                            1                     0                   2   
1                            1                     0                   2   
2                            2                     1                   0   

  weather_condition   latitude   longitude road_description slope_description  
0             clear  14.959105  100.873463    straight road          no slope  
1             clear  15.210738  104.862689    straight road          no slope  
2             clear  12.374259   99.907949       wide curve        slope area

### **Agency Column**

In [72]:
## Getting the categories of the column
data['agency'].unique()

array(['department of rural roads', 'department of highways',
       'expressway authority of thailand'], dtype=object)

We have three government agency involved :

*   department of rural roads  
*   department of highways
*   expressway authority of thailand




### **Route Column**

In [73]:
data['route'].head()

0    แยกทางหลวงหมายเลข 21 (กม.ที่ 31+000) - บ้านวัง...
1                                เดชอุดม - อุบลราชธานี
2                                ปราณบุรี - ปากน้ำปราณ
3    เชื่อมทางหลวงหมายเลข 1013 (กม.ที่ 8+200) - บ้า...
4                                     เกยไชย - ศรีมงคล
Name: route, dtype: object

From the list above, we noticed the route was recorded in thai language. so we will translate the language from thai to english.

In [74]:
# getting the unique route in the coulmn
data['route'].nunique()

3882

there are 3882 unique route in the column, we will translate this unique routes to english and use it as a dictionary to interprete the entire route column

In [75]:
# importing the csv file that has the translation of the route unique value
route_dic = pd.read_csv('https://raw.githubusercontent.com/LegendSeyi/Dataset/main/tranlated_thai_routes.csv', encoding='utf-8')
route_dic

Thai  \
0     แยกทางหลวงหมายเลข 21 (กม.ที่ 31+000) - บ้านวัง...   
1                                 เดชอุดม - อุบลราชธานี   
2                                 ปราณบุรี - ปากน้ำปราณ   
3     เชื่อมทางหลวงหมายเลข 1013 (กม.ที่ 8+200) - บ้า...   
4                                      เกยไชย - ศรีมงคล   
...                                                 ...   
3877  แยกทางหลวงหมายเลข 3204 (กม.ที่ 16+940) - บ้านขลู่   
3878  แยกทางหลวงหมายเลข 3 (กม.ที่ 452+700) - บ้านตาหนึก   
3879  แยกทางหลวง 3329 (กม.ที่ 16+720) -บ้านนิคมเขาบ่...   
3880    แยกทางหลวงหมายเลข 347 (กม.ที่ 21+700) - บ้านรุน   
3881  แยกทางหลวงหมายเลข 35 (กม.ที่ 76+070) - บ้านสี่แยก   

                                                English  
0             Highway 21 (km 31+000) - Ban Wang Phloeng  
1                           Dej Udom - Ubon Ratchathani  
2                               Pranburi - Pak Nam Pran  
3      Connect Highway 1013 (km 8+200) - Ban Khun Klang  
4                                Kayachai -Sri Mongkhon  
...                                                 ...  
3877  Highway intersection number 3204 (km 16+940) -...  
3878                Highway 3 (Km. 452+700) - Ban Ta Nu  
3879  Highway 3329 (Km 16+720) -Ban Nikhom Khao Bo Kaeo  
3880  Highway intersection number 347 (Km. 21+700) -...  
3881          Highway 35 (km 76+070) - Ban intersection  

[3882 rows x 2 columns]

In [76]:

# Create a mapping dictionary from the CSV file
mapping_dict = dict(zip(route_dic['Thai'], route_dic['English']))

data['route'] = data['route'].map(mapping_dict)
data.head()

acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   
3    571924    2019-01-01      00:20:00  2019-01-02    05:12:00   
4    599523    2019-01-01      00:25:00  2019-01-04    09:42:00   

              province                     agency  \
0               Loburi  department of rural roads   
1     Ubon Ratchathani     department of highways   
2  Prachuap Khiri Khan     department of highways   
3           Chiang Mai  department of rural roads   
4         Nakhon Sawan     department of highways   

                                              route           vehicle_type  \
0         Highway 21 (km 31+000) - Ban Wang Phloeng             motorcycle   
1                       Dej Udom - Ubon Ratchathani  private/passenger car   
2                           Pranburi - Pak Nam Pran             motorcycle   
3  Connect Highway 1013 (km 8+200) - Ban Khun Klang             motorcycle   
4                            Kayachai -Sri Mongkhon  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

In [77]:
data['route'].nunique()

3872

In [78]:
#Checking if the change is applied
data.head(3)

acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   

              province                     agency  \
0               Loburi  department of rural roads   
1     Ubon Ratchathani     department of highways   
2  Prachuap Khiri Khan     department of highways   

                                       route           vehicle_type  \
0  Highway 21 (km 31+000) - Ban Wang Phloeng             motorcycle   
1                Dej Udom - Ubon Ratchathani  private/passenger car   
2                    Pranburi - Pak Nam Pran             motorcycle   

                           presumed_cause                       accident_type  \
0  driving under the influence of alcohol                               other   
1                                speeding    rollover/fallen on straight road   
2                                speeding  head-on collision (not overtaking)   

   number_of_vehicles_involved  number_of_fatalities  number_of_injuries  \
0                            1                     0                   2   
1                            1                     0                   2   
2                            2                     1                   0   

  weather_condition   latitude   longitude road_description slope_description  
0             clear  14.959105  100.873463    straight road          no slope  
1             clear  15.210738  104.862689    straight road          no slope  
2             clear  12.374259   99.907949       wide curve        slope area

### **Vehicle Type Column**

In [79]:
data['vehicle_type'].head()

0               motorcycle
1    private/passenger car
2               motorcycle
3               motorcycle
4    private/passenger car
Name: vehicle_type, dtype: object

In [80]:
## checking if there is no null value
data['vehicle_type'].isnull().sum()

0

we have no null values

In [81]:
# Checking the catergory of vehicle involved in the accident
data['vehicle_type'].nunique()

15

15 different vehicle type was involved

### **Presumed Cause Column**

In [82]:
data['presumed_cause'].head()

0           driving under the influence of alcohol
1                                         speeding
2                                         speeding
3           driving under the influence of alcohol
4    cutting in closely by people/vehicles/animals
Name: presumed_cause, dtype: object

In [83]:
#checking if we have null values
data['presumed_cause'].isnull().sum()

0

**Checking the presumed cause of accident uniqueness**

In [84]:
data['presumed_cause'].unique()

array(['driving under the influence of alcohol', 'speeding',
       'cutting in closely by people/vehicles/animals',
       'failure to yield right of way', 'failure to yield/signal',
       'falling asleep', 'running red lights/traffic signals', 'other',
       'unfamiliarity with the route/unskilled driving',
       'vehicle equipment failure', 'illegal overtaking', 'tailgating',
       'ignoring stop sign while leaving intersection',
       'overloaded vehicle', 'insufficient light',
       'disabled vehicle without proper signals', 'abrupt lane change',
       'debris/obstruction on the road', 'reversing vehicle',
       'brake/anti-lock brake system failure', 'medical condition',
       'vehicle electrical system failure', 'driving in the wrong lane',
       'straddling lanes', 'dangerous curve',
       'failure to signal enter/exit parking', 'slippery road',
       'no traffic signs', 'sudden stop',
       'using mobile phone while driving',
       'driving without headlights/ill

In [85]:
data['presumed_cause'].nunique()

54

**We have 54 unique presumed cause of accident**

**From the above list, we can observe that some causes are not properly represented. The text format is not well represented in english language. so it is best to change it to english**

**We have 3 unique presumed causes that are not well represented**

In [86]:
# Using the googletrans library to translate the Thai text to english
!pip install googletrans==4.0.0-rc1

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.1/55.1 kB 2.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.4/133.4 kB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.6/42.6 kB 4.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.6/53.6 kB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.0/65.0 kB 4.8 MB/s eta 0:00:00
  Created wheel for googletrans: filename=googletrans-4.0.0rc1-py3-none-any.whl size=17396 sha256=3356a4cd7c132516a4a9e6e74b93f678bc5faee9186e384711e0c2d876122f5f
  Stored in directory: /root/.cache/pip/wheels/c0/59/9f/7372f0cf70160fe61b528532e1a7c8498c4becd6bcffb022de
Successfully built googletrans
  Attempting uninstall: chardet
    Found existing installation: chardet 5.2.0
    Uninstalling cha

In [87]:
#importing the translator from the googletrans library.
from googletrans import Translator

# translate the language texts to english
thai_values = ['ป้ายจราจรชำรุด', 'เส้นแบ่งทิศทางจราจรชำรุด', 'มึนเมาจากแอลกอฮอล์'
]

# Initialize the translator
translator = Translator()

# Translate Thai values to English and store in a dictionary
translated_dict = {}
for thai_value in thai_values:
    translation = translator.translate(thai_value, src='th', dest='en')
    translated_dict[thai_value] = translation.text

# Print the translated dictionary
for thai_value, english_translation in translated_dict.items():
    print(f"'{thai_value}': '{english_translation}'")

'ป้ายจราจรชำรุด': 'Damaged traffic signs'
'เส้นแบ่งทิศทางจราจรชำรุด': 'Damaged traffic direction'
'มึนเมาจากแอลกอฮอล์': 'Dizzy from alcohol'


In [88]:
# This the dictionary containing the thai and english text translation
translated_dict

{'ป้ายจราจรชำรุด': 'Damaged traffic signs',
 'เส้นแบ่งทิศทางจราจรชำรุด': 'Damaged traffic direction',
 'มึนเมาจากแอลกอฮอล์': 'Dizzy from alcohol'}

**Defining a function can will replace the thai text with English text**

In [89]:
# define function that will replace the texts with english
def change_lang(x):
  if x == 'ป้ายจราจรชำรุด':
    x = 'Damaged traffic signs'
    return x
  elif x == 'เส้นแบ่งทิศทางจราจรชำรุด':
    x ='Damaged traffic direction'
    return x
  elif x == 'มึนเมาจากแอลกอฮอล์':
    x ='Dizzy from alcohol'
    return x
  else:
    return x

**Applying the function to the presumed cause column**

In [90]:
data['presumed_cause'] = data['presumed_cause'].apply(change_lang)

In [91]:
# Checking if the change is applied
data['presumed_cause'].unique()

array(['driving under the influence of alcohol', 'speeding',
       'cutting in closely by people/vehicles/animals',
       'failure to yield right of way', 'failure to yield/signal',
       'falling asleep', 'running red lights/traffic signals', 'other',
       'unfamiliarity with the route/unskilled driving',
       'vehicle equipment failure', 'illegal overtaking', 'tailgating',
       'ignoring stop sign while leaving intersection',
       'overloaded vehicle', 'insufficient light',
       'disabled vehicle without proper signals', 'abrupt lane change',
       'debris/obstruction on the road', 'reversing vehicle',
       'brake/anti-lock brake system failure', 'medical condition',
       'vehicle electrical system failure', 'driving in the wrong lane',
       'straddling lanes', 'dangerous curve',
       'failure to signal enter/exit parking', 'slippery road',
       'no traffic signs', 'sudden stop',
       'using mobile phone while driving',
       'driving without headlights/ill

**Done**

### **Accident Type Column**

**Checking the categories or type of accident**

In [92]:
data.accident_type.unique()

array(['other', 'rollover/fallen on straight road',
       'head-on collision (not overtaking)',
       'collision at intersection corner',
       'collision with obstruction (on road surface)',
       'rear-end collision', 'pedestrian collision',
       'rollover/fallen on curved road', 'collision during overtaking',
       'turning/retreating collision', 'side collision'], dtype=object)

In [93]:
# Checking if there are null values
data.accident_type.isnull().sum()

0

**There are no null values**

### **Number of vehicles involved Column**

In [94]:
data.number_of_vehicles_involved.unique()

array([ 1,  2,  3,  0,  4,  6,  5,  7, 12, 11, 10,  8,  9, 14, 19, 27, 24,
       13])

**range is from 0 - 27 vehicles**

In [95]:
data.number_of_vehicles_involved.isnull().sum()

0

**Also we have no null value**

### **number of fatalities Column**

In [96]:
#Checking for null values
data.number_of_fatalities.isnull().sum()

0

In [97]:
#checking for the range of number of fatalities
data.number_of_fatalities.unique()

array([ 0,  1,  3,  2,  5,  6,  9,  4, 10,  7, 11, 13,  8])

### **Number of injuries Column**

In [98]:
# Checking if there is a null value
data.number_of_injuries.isnull().sum()

0

In [99]:
#Checking the range
data.number_of_injuries.unique()

array([ 2,  0,  1,  5,  4,  3,  9,  7, 12, 14,  6,  8, 51, 13, 10, 11, 17,
       15, 35, 24, 20, 21, 30, 19, 28, 32, 18, 25, 22, 16, 23, 27, 43, 42,
       46, 31, 39, 49, 38, 44, 34])

### **Weather condition column**

Checking for the types of weather condition

In [100]:
data.weather_condition.unique()

array(['clear', 'foggy', 'dark', 'rainy', 'other', 'natural disaster',
       'land slide'], dtype=object)

In [101]:
# Checking if there are null values
data.weather_condition.isnull().sum()

0

### **Latitude and Longitude Column**

**Checking if there are null value in the latitude and longitude column**

In [102]:
data[['latitude', 'longitude']].isnull().sum()

latitude     359
longitude    359
dtype: int64

**And yes, there are null values in the columns**

**There are 359 in the columns**

In [103]:
# Getting the count of the data will the missing information on latitide and longitude
data.loc[(data['latitude'].isnull()) & (data['latitude'].isnull())].count()

acc_code                       359
Incident_date                  359
Incident_time                  359
report_date                    359
report_time                    359
province                       359
agency                         359
route                          359
vehicle_type                   359
presumed_cause                 359
accident_type                  359
number_of_vehicles_involved    359
number_of_fatalities           359
number_of_injuries             359
weather_condition              359
latitude                         0
longitude                        0
road_description               359
slope_description              359
dtype: int64

In [104]:
long_lat_null = data[['province','route','latitude', 'longitude']].loc[data['latitude'].isnull()]

In [105]:
long_lat_null['route'].describe()

count                            359
unique                            22
top       Chaeng Watthana-Phaya Thai
freq                              64
Name: route, dtype: object

**Chaeng Watthana-Phaya Thai** is the most frequently occuring route with Null value on the lat and long column occurring **64** times

Also we have **22** unique routes with missing value of the long and lat

Due to fact that the analysis is not focused on geographical location, dropping the longitude and latitude column will be ideal. And also because they contain missing values

In [106]:
data.drop('longitude', axis=1, inplace=True)
data.drop('latitude', axis=1, inplace=True)

In [107]:
data.head(3)

acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   

              province                     agency  \
0               Loburi  department of rural roads   
1     Ubon Ratchathani     department of highways   
2  Prachuap Khiri Khan     department of highways   

                                       route           vehicle_type  \
0  Highway 21 (km 31+000) - Ban Wang Phloeng             motorcycle   
1                Dej Udom - Ubon Ratchathani  private/passenger car   
2                    Pranburi - Pak Nam Pran             motorcycle   

                           presumed_cause                       accident_type  \
0  driving under the influence of alcohol                               other   
1                                speeding    rollover/fallen on straight road   
2                                speeding  head-on collision (not overtaking)   

   number_of_vehicles_involved  number_of_fatalities  number_of_injuries  \
0                            1                     0                   2   
1                            1                     0                   2   
2                            2                     1                   0   

  weather_condition road_description slope_description  
0             clear    straight road          no slope  
1             clear    straight road          no slope  
2             clear       wide curve        slope area

**Done**

### **Road Description Column**

Checking the unique descriptions of the road

In [108]:
data['road_description'].unique()

array(['straight road', 'wide curve', 'other',
       'connecting to public/commercial area', 'sharp curve',
       'four-way intersection', 'connecting to private area',
       't-intersection', 'y-intersection',
       'grade-separated intersection/ramps', 'merge lane',
       'connecting to school area', 'lane-changing area', 'u-turn area',
       'roundabout', 'motorcycle lane', 'pedestrian path',
       'bridge (across river/canal)',
       'zebra crossing/pedestrian crossing'], dtype=object)

In [109]:
# how many unique description do we have
data['road_description'].nunique()

19

We have **19** unique description of the road

In [110]:
data['road_description'].isnull().sum()

0

There are no null values

**Done**

### **Slope Description Column**

Checking the unique descriptions of the slope

In [112]:
data['slope_description'].unique()

array(['no slope', 'slope area', 'other'], dtype=object)

There just **3** main categories for the slope description

In [113]:
data['slope_description'].isnull().sum(b)

0

There are no null Values in the column.


**Done**

## **EXPLORATORY DATA ANALYSIS (EDA)**



**In this section, well be analysing and exploring each column of the dataset as well as the insight we can derive from the dataset**

## **DATA VISUALIZATION**